src/lib/communication/DY.Lib.Communication.Core.Invariants.fst

```diff
@@ -0,0 +1,188 @@
+let pke_crypto_predicates_communication_layer #cusages = {
+  pred = (fun tr sk_usage pk msg ->
+    (exists sender receiver.
+      sk_usage == long_term_key_type_to_usage (LongTermPkeKey comm_layer_pkenc_tag)  receiver /\
+      (get_label tr msg) `can_flow tr` (join (principal_label sender) (principal_label receiver)) /\
+      event_triggered tr sender (CommConfSendMsg sender receiver msg)
+    )
+    );
+  pred_later = (fun tr1 tr2 sk_usage pk msg -> ());
+}
+
+val pke_crypto_predicates_communication_layer_and_tag: 
+  {|cusages:crypto_usages|} ->
+  (string & pke_crypto_predicate)
+let pke_crypto_predicates_communication_layer_and_tag #cusages = 
+  (comm_layer_pkenc_tag, pke_crypto_predicates_communication_layer)
+
+(*** Sign Predicates ***)
+
+#push-options "--ifuel 3 --fuel 0"
+val sign_crypto_predicates_communication_layer: {|cusages:crypto_usages|} -> sign_crypto_predicate
+let sign_crypto_predicates_communication_layer #cusages = {
+  pred = (fun tr sk_usage vk sig_msg ->
+    (match parse signature_input sig_msg with
+    | Some (Plain {sender; receiver; payload}) -> (
+      sk_usage == long_term_key_type_to_usage (LongTermSigKey comm_layer_sign_tag) sender /\
+      get_label tr payload `can_flow tr` public /\
+      event_triggered tr sender (CommAuthSendMsg sender payload)
+    )
+    | Some (Encrypted pk_receiver cm) -> (  
+      get_label tr cm.payload `can_flow tr` public /\
+      (exists plain_payload nonce.
+        sk_usage == long_term_key_type_to_usage (LongTermSigKey comm_layer_sign_tag) cm.sender /\
+        cm.payload == encrypt_message pk_receiver nonce plain_payload /\
+        event_triggered tr cm.sender (CommConfAuthSendMsg cm.sender cm.receiver plain_payload)
+      )
+    )
+    | None -> False)
+  );
+  pred_later = (fun tr1 tr2 sk_usage vk msg -> parse_wf_lemma signature_input (bytes_well_formed tr1) msg);
+}
+#pop-options
+
+val sign_crypto_predicates_communication_layer_and_tag: 
+  {|cusages:crypto_usages|} ->
+  (string & sign_crypto_predicate)
+let sign_crypto_predicates_communication_layer_and_tag #cusages =
+  (comm_layer_sign_tag, sign_crypto_predicates_communication_layer)
+
+val has_communication_layer_crypto_predicates:
+  {|crypto_invariants|} ->
+  prop
+let has_communication_layer_crypto_predicates #cinvs =
+  has_pke_predicate pke_crypto_predicates_communication_layer_and_tag /\
+  has_sign_predicate sign_crypto_predicates_communication_layer_and_tag
+
+(*** Event Predicates ***)
+
+noeq
+type comm_higher_layer_event_preds = {
+  send_conf: tr:trace -> sender:principal -> receiver:principal -> payload:bytes -> prop;
+  send_conf_later: 
+    tr1:trace -> tr2:trace ->
+    sender:principal -> receiver:principal -> payload:bytes ->
+    Lemma
+    (requires
+      send_conf tr1 sender receiver payload /\
+      bytes_well_formed tr1 payload /\
+      tr1 <$ tr2
+    )
+    (ensures send_conf tr2 sender receiver payload)
+  ;
```

Comment:
General thought: Does it ever make sense to have a send predicate over non well-formed payloads? This could be either added in the type of `send_conf`, or as a lemma similar to, e.g., `pred_knowable` for state predicates, from https://github.com/REPROSEC/dolev-yao-star-extrinsic/blob/182fd90a5284ab14ab18609c50397abba6581fa4/src/core/DY.Core.Trace.Invariant.fst#L42 .

Applies also to the other predicates here.